src/transformers/data/data_collator.py

```diff
@@ -133,3 +133,111 @@ def mask_tokens(self, inputs: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]
 
         # The rest of the time (10% of the time) we keep the masked input tokens unchanged
         return inputs, labels
+
+
+@dataclass
+class DataCollatorForXLNetLanguageModeling:
+    """
+    Data collator used for language modeling with XLNet.
+    - collates batches of tensors, honoring their tokenizer's pad_token
+    - preprocesses batches for masked language modeling with procedures specific to XLNet
+    """
+
+    tokenizer: PreTrainedTokenizer
+    mlm: bool = True
+    mlm_probability: float = 0.15
+    max_gram: int = 5  # maximum individual mask span length L
+
+    def __call__(self, examples: List[torch.Tensor]) -> Dict[str, torch.Tensor]:
+        batch = self._tensorize_batch(examples)
+        if self.mlm:
+            inputs, perm_mask, target_mapping, labels = self.mask_tokens(batch)
+            return {"input_ids": inputs, "perm_mask": perm_mask, "target_mapping": target_mapping, "labels": labels}
```

Comment:
We will have to pass the `mems` tensors here. Not sure yet what will be the best way to integrate this with the TF Trainer. Also pinging @TevenLeScao since he has a lot of experience training XLNet and Transfo-XL.